cpp/src/arrow/type.h

```diff
@@ -467,16 +486,20 @@ struct ARROW_EXPORT TimestampType : public FixedWidthType {
 
   int bit_width() const override { return static_cast<int>(sizeof(int64_t) * 8); }
 
-  TimeUnit unit;
-
   explicit TimestampType(TimeUnit unit = TimeUnit::MILLI)
       : FixedWidthType(Type::TIMESTAMP), unit(unit) {}
 
+  explicit TimestampType(const std::string& timezone, TimeUnit unit = TimeUnit::MILLI)
```

Comment:
We haven't resolved any questions about storing time zones in the IPC metadata (https://github.com/apache/arrow/blob/master/format/Message.fbs), so I did the bare minimum here to support the pandas and R use cases, where they have the notion of tz-naive and tz-aware timestamps. So I don't want to invest any more energy in it until we have a broader discussion about time zones